### ISSUE TITLE
Struct extension doesn't work with nested structs

### ISSUE THREAD

**Issue by paul** [dry-rb/dry-schema#464]
## Describe the bug

When generating a Dry::Schema from a Struct using the `:struct` extension, it first converts the input nested Hash into the Struct, then errors that the input isn't the expected hash.

## To Reproduce

I added a test case to `extensions/struct_spec.rb`:

```ruby
    context "complex struct" do
      let(:struct) do
        Class.new(Dry::Struct) do
          attribute :name, Types::String
          attribute :addresses, Types::Array do
            attribute  :city,   Types::String
            attribute? :street, Types::String
          end
        end
      end

      context "with valid input" do
        let(:input) do
          {foo: {name: "Jane", addresses: [{city: "New York"}]}}
        end

        it "has no errors" do
          expect(schema.(input).errors.to_h).to eq({})
        end
      end
```

```
Failures:

  1) struct extension value macro complex struct with valid input has no errors
     Failure/Error: expect(schema.(input).errors.to_h).to eq({})

       expected: {}
            got: {:foo=>{:addresses=>{0=>["must be a hash"]}}}

       (compared using ==)

       Diff:
       @@ -1 +1,2 @@
       +:foo => {:addresses=>{0=>["must be a hash"]}},

     # ./spec/extensions/struct_spec.rb:84:in `block (5 levels) in <top (required)>'
```

## Expected behavior

I'd expect this input to have no errors, since the nested hash satisfies the inner struct's schema. It appears something is converting the Hash into the given Struct, then errors because now that Struct isn't a Hash like its expecting. 

I spent some time digging into dry-types and dry-logic, but was unable to pin down exactly where it was attempting to verify the type of the struct was a Hash.

## My environment

- Ruby version: 3.2

**Comment by paul** [comment]
Update: I can make it work by monkey-patching Dry::Struct like this:

```ruby
module Dry
  class Struct
    def is_a?(klass)
      return true if klass == ::Hash

      super
    end

    def key?(name)
      @attributes.key?(name)
    end
  end
end
```

**Comment by paul** [comment]
Nope, nevermind, that one does weird things when the input is an empty array. It converts the empty array to an empty hash, then complains that the hash is not an array.

**Comment by flash-gordon** [comment]
Thanks for reporting! Can you try this branch out https://github.com/dry-rb/dry-schema/pull/466? It should work

**Comment by paul** [comment]
@flash-gordon That seems to work for me! It also works for the empty array situation in my previous commit. Should we add that as a test, too?

**Comment by flash-gordon** [comment]
@paul thanks, I did that. I'll merge the PR soon and cut a release with the fix

**Comment by flash-gordon** [comment]
Fixed in 1.13.3
